Add evalpoly function

MasonProtter opened the pull request:
Code taken from Steven G. Johnson's Juliacon 2019 talk. Gives same performance as `@horner`. 

```julia
horner(x) = zero(x)
horner(x, p1) = p1
horner(x, p1, ps...) = muladd(x, horner(x, ps...), p1)

x = 10

f(x) = horner(x, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0)
g(x) = Base.Math.@horner(x, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0)

@btime f($(Ref(x))[]) #  8.921 ns (0 allocations: 0 bytes)
@btime g($(Ref(x))[]) #  8.921 ns (0 allocations: 0 bytes)
```

_____
A lot has changed since I wrote this PR originally and the above code is not the current implementation. 

See https://github.com/JuliaLang/julia/pull/32753#issuecomment-554676094 for the current state of this PR.

MasonProtter committed:
implemented @big macro

MasonProtter committed:
added News entry

MasonProtter committed:
remove whitespace

MasonProtter committed:
added tests

MasonProtter committed:
Add horner function

MasonProtter committed:
Add horner function

MasonProtter committed:
remove whitespace

MasonProtter committed:
remove whitespace

KristofferC commented:
What's the point of having both?

MasonProtter committed:
Add tests to make sure @horner is the same as horner

MasonProtter committed:
Add tests to make sure @horner is the same as horner

MasonProtter commented:
I'd say there's no reason to have the macro if the function can do the same job just as fast, but the macro can't be removed until 2.0.

KristofferC commented:
Horners method is purely a syntactical transformation which is why it works so well as a macro. Hoping that the compiler will end up going through 12 calls of inlining (which might be true for floats but not for other types etc) seems strictly worse.

MasonProtter commented:
Good points! I guess my thinking was just that macros tend to be a bit more brittle than functions. One thing I like about this is that I can do things like
```
horner(x, somevector...)
```
I know that's not going to produce optimal code unless `somevector` is a compile time constant, but it's at least more flexible than `@horner`.

MasonProtter committed:
update horner function doc.

MasonProtter committed:
update horner function doc.

macd commented:
Unless I've done this wrong (which is always possible) it looks like the ability to splat into myhorner is a mis-feature?
```
julia v1.3> using BenchmarkTools

julia v1.3> function fhorner(x, coefs)
                b = coefs[end]
                for i in length(coefs)-1:-1:1
                    b = coefs[i] + x * b
                end
                return b
            end
fhorner (generic function with 1 method)

julia v1.3> myhorner(x) = zero(x)
myhorner (generic function with 1 method)

julia v1.3> myhorner(x, p1) = p1
myhorner (generic function with 2 methods)

julia v1.3> myhorner(x, p1, ps...) = muladd(x, myhorner(x, ps...), p1)
myhorner (generic function with 3 methods)

julia v1.3> x = 10
10

julia v1.3> y = [1.0, 2.0, 3.0]
3-element Array{Float64,1}:
 1.0
 2.0
 3.0

julia v1.3> @btime fhorner($x, $y)
  9.183 ns (0 allocations: 0 bytes)
321.0

julia v1.3> @btime myhorner($x, $(y)...)
  608.847 ns (14 allocations: 240 bytes)
321.0

julia v1.3> 
```

MasonProtter commented:
Splatting into `horner` will only be fast if the length of the container is known at compile time. For instance with tuples.
```julia
horner(x) = zero(x)
horner(x, p1) = p1
horner(x, p1, ps...) = muladd(x, horner(x, ps...), p1)

f(x, v) = horner(x, v...)
g(x, p1, p2, p3) = horner(x, p1, p2, p3)

v = (1.0, 2.0, 3.0)

@btime f($(Ref(x))[], $v)            # 1.696 ns (0 allocations: 0 bytes)
@btime g($(Ref(x))[], 1.0, 2.0, 3.0) # 1.696 ns (0 allocations: 0 bytes)
```

MasonProtter commented:
Another advantage of doing this as a function is that we could have `horner(x, v::Vector)` use the loop definition and `horner(x, tup::Tuple)` use the recursive inlining definition. Then there's a pretty simple API to switch between the definitions as needed.

macd commented:
Now that makes good sense.

MasonProtter committed:
whitespace fixes

MasonProtter committed:
whitespace fixes

MasonProtter committed:
More whitespace fixes

MasonProtter committed:
More whitespace fixes

stevengj commented:
If we add a function here, I would call it `evalpoly` rather than `horner`, and employ the Knuth algorithm for `evalpoly(x::Complex, ...)` similar to our `@evalpoly` macro.

See also #18410.

StefanKarpinski commented:
I'm not sure I see the advantage to having functions for `horner` and `evalpoly`. The macros guarantee that the optimizations you want are done since they directly generate that code. While it's cool that the compiler is smart enough to do the same optimization, why force it to? Why not just use the macro? Are there the cases where the function is preferable?

MasonProtter commented:
One case I showed above where the functions actually do something that the macros can’t do is stuff like
```
f(x, v::SVector) = evalpoly(1, v...)
```
Which is just as efficient as the macro. The crucial point here is that for `horner` to generate efficient code, I don’t actually need to know the polynomial coefficients ahead of time, I only need to know how many polynomial coefficients there should be. 

If we’re worried about having to rely on inlining to generate efficient code, we could also just do this as a generated function. Then we get to guarantee that the loop is unrolled but have (most of) the advantages of functions like using type level information to dispatch and generate custom implementations.

Commit:
Merge https://github.com/JuliaLang/julia into patch-2

MasonProtter committed:
replace horner func with evalpoly on real and complex

MasonProtter committed:
replace horner func with evalpoly on real and complex

MasonProtter committed:
Merge https://github.com/JuliaLang/julia into patch-2

MasonProtter committed:
added news and export statements

MasonProtter committed:
added news and export statements

MasonProtter commented:
Hi all, sorry for the hiatus on this, I was away on vacation. I have made some overhauls:
* Renamed the function `horner -> evalpoly`.
* Switched from relying on constant propagation / inlining to using a generated function.
* Implemented the Goertzel-like algorithm as a method: `evalpoly(z::Complex, p...)`.

I also thought a bit more about why this makes sense as a function more than a macro. To clarify what I said before, the loop unrolling being done in `@evalpoly` really only relies on one piece of information: the number of polynomial coefficients. One could say that it is doing a syntax transformation, but I think it's actually more accurate to say that it is using type-level information: the length of the tuple of coeffs. In that sense, this functionality is much more the domain of generated functions. 

In the `@evalpoly` source code, it uses an `if-else` to decide if `z` is real (in which case it uses horners method`) or complex in which case it uses other code, so it kindof already has a sort of fake internal dispatch going on inside. It's much nicer to use julia's built in dispatch and that way packages can extend it, like for instance if someone found a nice algorithm for evaluating quaternionic polynomials, it could easily be a method of `evalpoly`, but not `@evalpoly`. 

_________

Looking forward, my `evalpoly(x, p...)` is currently implemented using `if @generated` so that the compiler may choose to use a slower, looping definition if it needs to, but I have not yet done the same for `evalpoly(z::Complex, p...)`. Would people prefer if `evalpoly(z::Complex, p...)` used `if @generated` or is a regular `@generated` function fine? 

Perhaps I should instead get rid of the option to not use the `@generated` version of `evalpoly(x, p...)`?

MasonProtter committed:
remove whitespace

MasonProtter committed:
remove whitespace

MasonProtter committed:
fix broken evalpoly(x, p...)

MasonProtter committed:
fix broken evalpoly(x, p...)

MasonProtter committed:
fix whitespace yet again

MasonProtter committed:
fix whitespace yet again

StefanKarpinski commented:
This is looking good! @stevengj, would you mind reviewing?

KristofferC commented:
This seems like a point of confusion:

```
julia> coffs = (1.0, 2.0, 3.0)
(1.0, 2.0, 3.0)

julia> evalpoly(2.0, coffs...)
17.0

julia> evalpoly(2.0, coffs)
(1.0, 2.0, 3.0)
```

martinholters commented:
Late to the party, but I think I might prefer taking the coefficients as one parameter (tuple) instead of one parameter per coefficient.

MasonProtter commented:
@KristofferC That is actually also reproduced in `@horner`: 
```
julia> Base.Math.@horner(2, (1,2,3))
(1, 2, 3)
```
The reason just comes down the the `muladd` logic. 

However, if we go with @martinholters suggestion, and use `evalpoly(x, coeffs)` to mean what `evalpoly(x, coeffs...)` currently means, that issue would be resolved. 

________

Looking into this lead me to find what might be considered a bug in the current `@evalpoly` implementation (and by extension `evalpoly(z::complex, p...)`): The Goertzel-like algorithm assumes there are two or more coefficients
```
julia> @evalpoly(2+im, 1, 2)
5 + 2im

julia> @evalpoly(2+im, 1)
ERROR: LoadError: BoundsError: attempt to access (1,)
  at index [0]
Stacktrace:
 [1] getindex(::Tuple, ::Int64) at ./tuple.jl:24
 [2] @evalpoly(::LineNumberNode, ::Module, ::Any, ::Vararg{Any,N} where N) at ./math.jl:184
in expression starting at REPL[9]:1
```

This can be easily fixed in the function form by adding a method to `evalpoly(z::Complex, p::Tuple{T}) where {T} = p[1]`, but in the macro will require a new `if-else`.

MasonProtter committed:
fix spurious [@]evalpoly(z, p) behaviour

MasonProtter committed:
fix spurious [@]evalpoly(z, p) behaviour

MasonProtter committed:
test fixed spurious [@]evalpoly(z, p) behaviour

MasonProtter committed:
test fixed spurious [@]evalpoly(z, p) behaviour

MasonProtter commented:
I took it upon myself to fix the `[@]evalpoly` errors in this PR.

musm commented:
I'm in favor as well. Should the existing `Base.Math.@horner` be deprecated or removed? (It's not exported so I assume it can be removed)

With the optionally-generated function implementation I assume the unrolled version is always used and the non-generated version is used for cases when a non-statically sized vector is passed?

DilumAluthge commented:
Why can’t we keep the `Base.Math.@horner` macro?

MasonProtter committed:
change evalpoly API to take a tuple of coefficients

MasonProtter committed:
change evalpoly API to take a tuple of coefficients

MasonProtter commented:
Seeing as @martinholters's suggestion got a thumbs up form @stevengj and @musm, I went ahead and tried it out. The appears to be a slight performance regression with this change which gets quite bad if the user supplies a non `NTuple` like `(1, 2, 3.0)` for the coefficients. I'll have to think a bit more about how to do this right, otherwise it might be best to revert to the `varargs` definition.

musm commented:
@MasonProtter  that looks like because   `p::NTuple{N}` in the method signature and since in your example the tuple elements don't all have the same type.  I'm not sure there's a good way to allow various types for the coefficients to be allowed  while still known the length of the tuple `N`. 


> Why can’t we keep the Base.Math.@horner macro?

Because `evalpoly` would do exactly the same and has more capability, so keeping `@horner` would be  redundant code sitting in Base.

MasonProtter committed:
fix performance regression due to promote

MasonProtter committed:
fix performance regression due to promote

MasonProtter committed:
Fix borked earlier commit

MasonProtter committed:
Fix borked earlier commit

musm commented on base/math.jl:
```diff
@@ -102,6 +102,77 @@ macro horner(x, p...)
+
+
+"""
+    evalpoly(z::Complex, p::NTuple{N})  where {N}
```
```suggestion
    evalpoly(z::Complex, p::Tuple)
```

musm commented on base/math.jl:
```diff
@@ -102,6 +102,77 @@ macro horner(x, p...)
 end
 
+"""
+    evalpoly(x, p::NTuple{N}) where {N}
```
```suggestion
    evalpoly(x, p::Tuple)
```

MasonProtter replied:
Blah. Good catches. I had two branches going and wasn't pushing all the changes.

MasonProtter committed:
fix docstrings

MasonProtter committed:
fix docstrings

MasonProtter committed:
fix docstrings again

MasonProtter committed:
fix docstrings again

musm commented:
I'm in favor of this change but we should really think about getting rid of `@evalpoly` and `@horner` since this essentially supersedes them. Perhaps this is better suited for a seperate PR.

MasonProtter commented:
I don't believe it would be appropriate to remove `@evalpoly` as it is exported. `@horner` should be okay to remove but perhaps it should give a deprecation warning instead of outright removing it? I'm pretty sure it is used somewhat commonly out there in the ecosystem.

stevengj commented:
Note that the `@evalpoly(x, c...)` macro could simply call `evalpoly(x, (c...,))`

stevengj commented on base/math.jl:
```diff
@@ -102,6 +102,77 @@ macro horner(x, p...)
+Evaluate the polynomial ``\\sum_k p[k] z^{k-1}`` for the coefficients `p[1]`, `p[2]`, ...;
+that is, the coefficients are given in ascending order by power of `z`. This function
+generates efficient code using a Goertzel-like algorithm specialized for complex arguments
+with loops unrolled at compile time.
```
Would be good to add the Knuth citation somewhere.

MasonProtter replied:
Done.

stevengj commented:
Note: closes #7186.

MasonProtter committed:
Cite Knuth

MasonProtter committed:
Cite Knuth

MasonProtter committed:
remove @horner

MasonProtter committed:
remove @horner

MasonProtter committed:
remove whistspace

MasonProtter committed:
remove whistspace

MasonProtter commented:
Hm. I'm not sure why, but removing `@horner` makes `make` bug out... Unless I did something else foolish. Reverting for now.

MasonProtter committed:
undo remove @horner

MasonProtter committed:
undo remove @horner

musm commented:
Note if removing `@horner`  you have to replace all instances of the `@horner` macro with `evalpoly`. There are several instances in the math library where it's being used.

musm commented:
We should probably open an issue to track removal of `@evalpoly` for future versions.

stevengj commented on base/math.jl:
```diff
@@ -142,6 +128,9 @@ that is, the coefficients are given in ascending order by power of `z`. This fun
 generates efficient code using a Goertzel-like algorithm specialized for complex arguments
 with loops unrolled at compile time.
 
+The Goertzel-like algorthim is described in Knuth's Art of computer programming, 
```
s/algorthim/algorithm, and fix titlecase in the book title

stevengj commented:
`@horner` is also used in SpecialFunctions.jl for historical reasons

MasonProtter committed:
fix spelling and ref in docstring

MasonProtter committed:
fix spelling and ref in docstring

MasonProtter committed:
fix spelling again

MasonProtter committed:
fix spelling again

MasonProtter committed:
Merge https://github.com/JuliaLang/julia into patch-2

MasonProtter committed:
Merge branch 'master' into patch-2

musm commented:
Can you rebase? Test failures are unrelated

MasonProtter committed:
Merge branch 'master' into patch-2

MasonProtter committed:
fix merge conflict

MasonProtter committed:
fix merge conflict

MasonProtter commented:
> Can you rebase? Test failures are unrelated

Did I do that right? Or do you want the commits squashed?

ViralBShah commented:
Squashing can be done during merging.

MasonProtter committed:
remove erroneous change from

MasonProtter committed:
Update util.jl

MasonProtter committed:
Remove erroneous change

musm commented on base/math.jl:
````diff
@@ -102,6 +102,80 @@ macro horner(x, p...)
+17
+```
+"""
+function evalpoly(x, p::Tuple)
````
shouldn't `z::Real`  (probably doesn't matter)?

MasonProtter replied:
It was intentional, since Horner's rule should apply to any object for which multiplication and addition of scalars are defined.

musm commented:
The more I think about it isn't the original API with the signature that takes in the coefficients better optimized for more generic cases, e.g.

```julia
function evalpoly1(x, p...)
    if @generated
        ex = :(p[end])
                for i in length(p)-1:-1:1
            ex = :(muladd(x, $ex, p[$i]))
        end
        ex
    else
        ex = p[end]
        for i in length(p)-1:-1:1
            ex = muladd(x, ex, p[i])
        end
        ex
    end
end

 function evalpoly2(x, p::Tuple)
    if @generated
        N = length(p.parameters)
        ex = :(p[end])
        for i in N-1:-1:1
            ex = :(muladd(x, $ex, p[$i]))
        end
        ex
    else
        N = length(p)
        ex = p[end]
        for i in N-1:-1:1
            ex = muladd(x, ex, p[i])
        end
        ex
    end
end


evalpoly1(x,a::Vector) = evalpoly1(x,a...)
evalpoly2(x,a::Vector) = evalpoly2(x,(a...,))

julia> @btime evalpoly1(1.0, [1.0,2.0,3.0])
  156.346 ns (6 allocations: 192 bytes)
6.0

julia> @btime evalpoly2(1.0, [1.0,2.0,3.0])
  380.198 ns (5 allocations: 192 bytes)
6.0
```

MasonProtter commented:
Anyone else have thoughts or feelings regarding whether the signature should be `evalpoly(x, p::Tuple)` or `evalpoly(x, p...)`? I don't have strong preferences either way. 

@musm's performance regression on `Vector`s could be solved by just making a method
```julia
function evalpoly(x, p::AbstractVector)
    N = length(p)
    ex = p[end]
    for i in N-1:-1:1
        ex = muladd(x, ex, p[i])
    end
    ex
end
```
and then one could just make a PR to `StaticArrays.jl` defining
```
evalpoly(x, p::SVector) = evalpoly(x, p.data)
```

musm commented:
@MasonProtter   that would be a good idea, if we decide upon   `evalpoly(x, ::Tuple)` method, which, to be clear, I'm not actually against.

MasonProtter commented:
Is this something that can make it into 1.4 before the freeze? I'd really appreciate it if I could get some feedback.

ViralBShah commented:
What is holding this back? I guess it at least needs a rebase.

JeffBezanson commented:
Triage hesitates to add this, especially since it's a generated function. Some thoughts:

- If we have both a macro and a function, we could say to use the macro if you need code generation, and otherwise you get the function, which would do a loop and handle any iterable.
- The code generator only needs to know the number of coefficients, so we might want to add a version of the macro that accepts that, e.g. `@evalpoly_n 5 c` which would generate code with `c[1]` through `c[5]` as the coeffs.

MasonProtter commented:
> Triage hesitates to add this, especially since it's a generated function

Why does triage not want new generated functions in Base? Is there some generated function edge case that makes this undesirable? Note that it's an optionally generated function, and it was my understanding that would alleviate at least some of the concerns. 

> The code generator only needs to know the number of coefficients, so we might want to add a version of the macro that accepts that, e.g. @evalpoly_n 5 c which would generate code with c[1] through c[5] as the coeffs.

I guess I just don't understand why one would want to do this with a macro in the first place when this is a textbook use-case for generated functions, and chunk of the `@evalpoly` macro is devoted to ['faking' dispatch](https://github.com/JuliaLang/julia/blob/148761425d448467eb97aab534b13ba877fb4982/base/math.jl#L224) which would be trivial and extendable with a function. 

I don't really see any reason to use a macro here other than the fact that we can't remove the macro until 2.0. It just seems strictly worse than a generated function as far as I can tell, so I don't see why we can't provide one whilst we wait to remove the macro, unless there's some pressing issue where the presence of a generated function would have unpleasant downstream effects?

JeffBezanson commented:
> and chunk of the `@evalpoly` macro is devoted to ['faking' dispatch](https://github.com/JuliaLang/julia/blob/148761425d448467eb97aab534b13ba877fb4982/base/math.jl#L224) which would be trivial and extendable with a function.

That's a fair point. I think I'd prefer it if we added a method that works on an iterable, which can also be used as the fallback for the non-generated case. That way at least you won't be tempted to write `evalpoly(x, Tuple(array))` which would be really slow. The Complex method also needs `if @generated`.

MasonProtter committed:
Merge branch 'master' into patch-2

MasonProtter committed:
add optional @generated and evalpoly(x, ::Vector)

MasonProtter committed:
test evalpoly(z, ::Vector)

MasonProtter commented:
> I think I'd prefer it if we added a method that works on an iterable, which can also be used as the fallback for the non-generated case. ... The Complex method also needs if @generated.

Done and done (assuming tests pass).

MasonProtter commented on base/math.jl:
````diff
@@ -95,6 +95,131 @@ macro horner(x, p...)
+17
+```
+"""
+evalpoly(x, p::AbstractVector) = _evalpoly(x, p)
````
I chose to restrict this to abstract vector, but not `_evalpoly(x, p)` (same for the complex case). I'm not sure if this is the preferred way to go about this, but it seemed like the most straightforward way to do this with minimal code duplication.

MasonProtter committed:
Make @horner and @evalpoly call the evalpoly function

MasonProtter committed:
fix whitespace

MasonProtter committed:
fix @evalpoly and @horner

MasonProtter committed:
fix @evalpoly and @horner more

MasonProtter committed:
typo

MasonProtter commented:
Okay, the tests now pass. Let me sum up this PR here for anyone interested in reviewing who doesn't want to go through all the comments above to figure out what's going on here: 

_____

### Short version: 

New function: `evalpoly` with methods
- `evalpoly(x::Any, p::Tuple)` which uses Horner's rule to calculate the polynomial `sum_k x^k * p[k+1])` efficiently with loops unrolled at compile time. This is of great use when we know the number of coefficients at compile time. This method does not suffer the [bug](https://github.com/JuliaLang/julia/pull/32753#issuecomment-529593230) that the current `@horner` suffers when given a single polynomial coefficient. This method is optionally `@generated`.
- `evalpoly(x::Any, p::AbstractVector)` which uses Horner's rule to calculate the polynomial `sum_k x^k * p[k+1])` efficiently *without* loops unrolled at compile time. This is useful for those who have a polynomial whose length is not statically known but still want to generate efficient code. This is a regular method. Not `@generated`. 
- `evalpoly(z::Complex, p::Tuple)` which uses a specialized algorithm for complex polynomials that's more efficient than Horner's rule to calculate the polynomial `sum_k z^k * p[k+1])` efficiently with loops unrolled at compile time. This is of great use when we know the number of coefficients at compile time. This method does not suffer the [bug](https://github.com/JuliaLang/julia/pull/32753#issuecomment-529593230) that the current `@evalpoly` suffers when given a single polynomial coefficient. This method is optionally `@generated`.
- `evalpoly(x::Complex, p::AbstractVector)` which uses the specialized complex polynomial algorithm without unrolling loops or knowing the number of coefficients until runtime.This is a regular method. Not `@generated`. 

Refactored macros
- `@horner` and `@evalpoly` now simply lower to calls to  `evalpoly`. This is to reduce code duplication while we wait for version 2.0 where they can be removed. 

_____

### Long version:

We currently have two macros: `@horner` and `@evalpoly` whose purpose is efficient computation of real and complex polynomials. `@evalpoly` is exported but `@horner` is not. `@horner` specializes on real numbers and if you give `@evalpoly` a real number, it will just lower to a call to `@horner`. The macros essentially do something like
```julia
@evalpoly(x, p1, p2, p3) => p1 + x*(p2 + x*(p3)))
```
for any number of coefficients `p1`, `p2`, and `p3`. This is more efficient than computing `p1 + x * p2 + x^2 * p3`. If `x` is a complex number, we use a more sophisticated algorithm that gives some additional speedup.

This was implemented as a macro because it seems kinda like a syntax transformation (they may have also been implemented before we had generated functions, I'm not actually sure about the history), but actually the only thing these macros really need to know to generate efficient code is the number of polynomial coefficients and the type of `x`. Using the `Vararg` syntax, we are able to tell the macros the number of coefficients, but not the type of `x` since variables don't have types at macro-expansion time, so the current implementation of `@evalpoly` just returns code that checks the type of `x` to decide what algorithm to use and actually generates the code for both (though the extra code is usually elided). This is what I call "faking dispatch". 

Modern julia has `@generated function`s whose purpose is to do specialized code generation based on the types of it's arguments. Hence, in this PR I made a function called `evalpoly` which takes two arguments: `x` the polynomial variable and `p`, an iterable collection of polynimial coefficients. In the case where `p` is a tuple, `evalpoly` uses a generated method to product the same code that `@evalpoly` or `@horner` would generate because `Tuple`s encode their length in their type. We can also do real™ dispatch on whether or not the variable is real or complex and now can support packages adding methods that might have specialized code for say, quaternions or clifford algebras, colour vectors spaces or whatever else. 

The generated methods also have fallback methods that are used if the compiler decides for whatever reason that doing the code generation is too burdensome, so it is free to instead do a runtime for loop which would be like
```julia
out = ps[1]
for p in ps[2:end]
    out += x*p
end
return out
```
We also dispatch to this runtime for loop definition in the case where a `Vector` of coefficients is passed to `evalpoly` so that users can get the algorithmic advantages of the implementation even if they don't know the number of coefficients at compile time.

to cut down code duplication until the macros are allowed to be removed, the macros now simply lower to `evalpoly` function calls. 

This PR also includes a fix to this current problem with `evalpoly` when given a single coefficient:
```julia
julia> @evalpoly(1, 1)
ERROR: LoadError: BoundsError: attempt to access (1,)
  at index [0]
Stacktrace:
 [1] getindex(::Tuple, ::Int64) at ./tuple.jl:24
 [2] @evalpoly(::LineNumberNode, ::Module, ::Any, ::Vararg{Any,N} where N) at ./math.jl:131
in expression starting at REPL[32]:1
```

MasonProtter committed:
Docfix

musm commented:
I'm in strong support of this PR

musm commented on base/math.jl:
```diff
@@ -81,18 +81,139 @@ function clamp!(x::AbstractArray, lo, hi)
-    end
-    ex = quote local r = $ex end # structure this to add exactly one line number node for the macro
-    return Expr(:block, :(local t = $(esc(x))), ex, :r)
+    esc(:(evalpoly($x, ($(p...),))))
```
shouldn't the `esc` only be over the arguments `x` and `p`. (Same for below.)

MasonProtter replied:
My understanding is that `esc`ing the entire expression is equivalent, but I could be wrong.

MasonProtter replied:
I guess one difference is that if I `esc` the entire expression, that relies on the user having `evalpoly` in their workspace, so this could give unexpected results if the user defines their own `evalpoly`. I'll change it.

MasonProtter replied:
Okay, I think I changed it correctly. I'll wait for the tests to pass before clicking resolve.

MasonProtter committed:
change `esc` logic

musm commented:
@MasonProtter

For very long polynomial coefficients, it turns out to be faster to use the array based version instead of the completely unrolled version. 

Therefore in my tests simply defining `evalpoly` using `_evalpoly` turns out to have better performance  for  long coefficient lists. (For short lists the compiler unrolls the loop and for long lists it uses the loop-based definition, which has better performance)
 
Thus I'm not convinced the `@generated` is entirely needed in the `::Tuple` case. Basically using `evalpoly` with a list in this PR **guarantees**  it is unrolled, whether this is the desired behavior or not is up for debate, since for very long coefficients the rolled version is faster typically by 5-10%.

In most cases as the tuple length is likely to have a length no more than say 20 coefficients, for which in

I'm not sure if you have an opinion on this.

MasonProtter commented:
@musm Wow! I hadn't realized that the compiler was *that* good at dealing with `Tuple`s. I had assumed it would be similar to the recursive inlining version but never checked. 
```julia
mylog1(x, ::Val{N}) where {N} = evalpoly(x - 1, ntuple(n -> (-1)^n/n, N))
mylog2(x, ::Val{N}) where {N} = Base.Math._evalpoly(x - 1, ntuple(n -> (-1)^n/n, N))

z = 1.0 + im
```
```julia
julia> @btime mylog1($(Ref(z))[], $(Val{5}())); @btime mylog2($(Ref(z))[], $(Val{5}()));
  18.582 ns (1 allocation: 48 bytes)
  17.691 ns (1 allocation: 48 bytes)

julia> @btime mylog1($(Ref(z))[], $(Val{20}())); @btime mylog2($(Ref(z))[], $(Val{20}()));
  1.126 μs (26 allocations: 864 bytes)
  1.122 μs (26 allocations: 864 bytes)

julia> @btime mylog1($(Ref(z))[], $(Val{50}())); @btime mylog2($(Ref(z))[], $(Val{50}()));
  2.337 μs (56 allocations: 1.83 KiB)
  2.337 μs (56 allocations: 1.83 KiB)

julia> @btime mylog1($(Ref(z))[], $(Val{100}())); @btime mylog2($(Ref(z))[], $(Val{100}()));
  4.266 μs (106 allocations: 3.36 KiB)
  4.266 μs (106 allocations: 3.36 KiB)

julia> @btime mylog1($(Ref(z))[], $(Val{200}())); @btime mylog2($(Ref(z))[], $(Val{200}()));
  8.189 μs (206 allocations: 6.61 KiB)
  8.178 μs (206 allocations: 6.61 KiB)

julia> @btime mylog1($(Ref(z))[], $(Val{500}())); @btime mylog2($(Ref(z))[], $(Val{500}()));
  24.508 μs (506 allocations: 16.00 KiB)
  24.231 μs (506 allocations: 16.00 KiB)
```
I think based on this, it may be safe to say that you're right and the `@generated` methods are unnecessary. I'd like to wait and and see if anyone, especially @JeffBezanson or @stevengj might have anything to say about their feelings towards forced unrolling in the generated function vs. just letting `Tuple` magic take care of it.

JeffreySarnoff commented:
This, in its most recent revision, is an improvement imo.

timholy commented:
Tuple-magic is also *considerably* faster to compile for short tuples (I can't remember the issue, but I think it was Kristoffer Carlsson who posted a nice analysis). You can automatically fall back to a loop with this trick: https://github.com/JuliaLang/julia/blob/master/base/tuple.jl#L156-L173.

stevengj commented:
(For long arrays, I think that `evalpoly` should switch over to a pairwise-like method that recomputes the `x^n` for sufficiently large base cases in order to improve accuracy, but that can be a separate PR.)

MasonProtter commented:
Do you have any thoughts on whether we should bother doing manual code generation here when the compiler seems to just do the right thing automatically?

chriselrod commented:
What about [Estrin's scheme](https://en.wikipedia.org/wiki/Estrin%27s_scheme) for evaluating polynomials with more coefficients? It can take advantage of out of order execution.

If the coefficients are known at compile time / memory layout of the coefficients is handled appropriately, it could even take advantage of SIMD.

Some of the higher coefficients could also be recomputed for accuracy, as stevengj suggested.

JeffreySarnoff commented:
A while back, I did some timings comparing Horner's method, Estrin's method, and a further parallelization that is available for some polys (I don't recall the name right now).  This was done before Julia's v1.3 multithreading.  I found that each was better than the others for some mixes of number of coeffs, poly degree, and type of coeffs.  The pairwise approach @stevenjg suggests probably is available in the most important part (keeping results accurate) in some results from the third method - which used software to develop, pare and repair computationally organized fixed length poly code. 

It seems that this is another circumstance where having access to a set of timings, to include some honed set of polynomial evaluations, generated at the time of Julia's installation / Julia version updating would be helpful.

MasonProtter commented:
I'd be happy to write a pair-wise algorithm. I'd have a slight preference to keep it in a separate PR though, since this one already has a somewhat long history and I anticipate a lot of bike-shedding and tweak suggestions if we get into algorithmic changes.

MasonProtter commented:
Okay, so I think I have identified one case where the tuple magic is inferior to generated functions: cases where the compiler knows *everything*. 
```julia
julia> function evalpoly1(x, p::Tuple)
           N = length(p)
           ex = p[end]
           for i in N-1:-1:1
               ex = muladd(x, ex, p[i])
           end
           ex
       end
evalpoly1 (generic function with 1 method)

julia> @generated function evalpoly2(x, p::Tuple)
           N = length(p.parameters)
           ex = :(p[end])
           for i in N-1:-1:1
               ex = :(muladd(x, $ex, p[$i]))
           end
           ex
       end

julia> @btime evalpoly1($(1.0), $(ntuple(n -> (-1)^n / n, 50)))
  39.777 ns (0 allocations: 0 bytes)
-0.6832471605759182

julia> @btime evalpoly2($(1.0), $(ntuple(n -> (-1)^n / n, 50)))
  0.019 ns (0 allocations: 0 bytes)
-0.6832471605759182

julia> f(x) = evalpoly1(x, (1.0, 2.0, 3.0, 4.0));  g(x) = evalpoly2(x, (1.0, 2.0, 3.0, 4.0));

julia> @btime f($(1.0))
  1.559 ns (0 allocations: 0 bytes)
10.0

julia> @btime g($(1.0))
  0.019 ns (0 allocations: 0 bytes)
10.0
```
So in cases where the compiler is able to statically know not only the coefficients but also the variable, the generated method will be completely hoisted but the regular method will not, likely due to the explicit loop.

yuyichao commented:
I very much doubt evalpoly1 can be vectorized in a loop. Ref https://github.com/JuliaLang/julia/issues/18410#issuecomment-245667179

JeffreySarnoff commented:
What remains to be _done_ _determined_ _decided_ for merging a consensus into v1.4?

musm commented:
@MasonProtter  your benchmark timings are meaningless since their sub `ns` to overcome the BenchmarkTools issue you need to do

```julia
@btime evalpoly1($(Ref(1.0))[], $(ntuple(n -> (-1)^n / n, 50)))
  48.683 ns (0 allocations: 0 bytes)
@btime evalpoly2($(Ref(1.0))[], $(ntuple(n -> (-1)^n / n, 50)))
  36.253 ns (0 allocations: 0 bytes)
```

In addition, code generation shows that `f` and `g` are identical.

MasonProtter committed:
Remove outdated test

musm commented:
@yuyichao  I'm not arguing for or against evalpoly1 or evalpoly2, but my tests show that identical code is generated in both cases using your test example in  https://github.com/JuliaLang/julia/issues/18410#issuecomment-245667179

musm commented:
@MasonProtter  upon closer thinking I'm in support of the PR as it currently stands.  From my perspective it's better to guarantee unrolling in the static tuple case.

MasonProtter commented:
> @MasonProtter your benchmark timings are meaningless since their sub `ns` to overcome the BenchmarkTools issue you need to do
> 
> ```julia
> @btime evalpoly1($(Ref(1.0))[], $(ntuple(n -> (-1)^n / n, 50)))
>   48.683 ns (0 allocations: 0 bytes)
> @btime evalpoly2($(Ref(1.0))[], $(ntuple(n -> (-1)^n / n, 50)))
>   36.253 ns (0 allocations: 0 bytes)
> ```
> 
> In addition, code generation shows that `f` and `g` are identical.

No, I don't believe they are meaningless. You're measuring the case where the compiler doesn't know `x` ahead of time. On the flip side, I was measuring the case where the compiler knows everything ahead of time and showed that the generated method was able to be completely hoisted out of the benchmark loop whereas the non-generated method wasn't. This can be advantageous in the cases where someone writes a completely static polynomial in their code like `evalpoly(2.0, (1.0, 2.0, 3.0))` because it'll just be replaced with the value. 

However, further testing I've done has shown that this process is fairly finicky and the difference in behaviours between the two methods is not always reliable. For instance, 
```julia
julia> f() = evalpoly1(1.0, ntuple(n -> (-1)^n / n, 50))
f (generic function with 1 method)

julia> g() = evalpoly2(1.0, ntuple(n -> (-1)^n / n, 50))
g (generic function with 1 method)

julia> @btime f()
  2.305 μs (54 allocations: 1.77 KiB)
-0.6832471605759182

julia> @btime g()
  2.291 μs (55 allocations: 1.78 KiB)
-0.6832471605759182
```
shows basically no difference, though I'd have naïvely hoped that julia should be able to just lower both of these functions to just `return -0.6832471605759182`. You're right of course that for large coefficient lists, it's valuable to drop to the looping definition. As @timholy  mentioned though, this can be done using that same `Any16` trick employed for `map( ::Any, ::Tuple)`.

musm commented:
I guess the only question remaining here is if we should employ the trick to drop to the looping definition after a threshold.

MasonProtter commented:
@musm I don't see a point where the non-generated method ever overtakes the generated method. Keep in mind that `evalpoly1` was the non-generated method here.  
```julia
julia> function evalpoly1(x, p::Tuple)
           N = length(p)
           ex = p[end]
           for i in N-1:-1:1
               ex = muladd(x, ex, p[i])
           end
           ex
       end
evalpoly1 (generic function with 1 method)

julia> @generated function evalpoly2(x, p::Tuple)
           N = length(p.parameters)
           ex = :(p[end])
           for i in N-1:-1:1
               ex = :(muladd(x, $ex, p[$i]))
           end
           ex
       end
evalpoly2 (generic function with 1 method)

julia> for N in [1, 10, 30, 100, 200, 500]
           @show N
           print("Non-generated: "); @btime evalpoly1($(Ref(1.0))[], $(ntuple(n -> (-1)^n / n, N)))
           print("Generated:     "); @btime evalpoly2($(Ref(1.0))[], $(ntuple(n -> (-1)^n / n, N)))
       end
N = 1
Non-generated:   0.029 ns (0 allocations: 0 bytes)
Generated:       0.029 ns (0 allocations: 0 bytes)
N = 10
Non-generated:   2.309 ns (0 allocations: 0 bytes)
Generated:       2.308 ns (0 allocations: 0 bytes)
N = 30
Non-generated:   18.260 ns (0 allocations: 0 bytes)
Generated:       12.494 ns (0 allocations: 0 bytes)
N = 100
Non-generated:   85.523 ns (0 allocations: 0 bytes)
Generated:       60.561 ns (0 allocations: 0 bytes)
N = 200
Non-generated:   196.920 ns (0 allocations: 0 bytes)
Generated:       162.365 ns (0 allocations: 0 bytes)
N = 500
Non-generated:   530.984 ns (0 allocations: 0 bytes)
Generated:       495.938 ns (0 allocations: 0 bytes)
```

musm commented:
Agreed. Ok I think we should go ahead and merge this PR as is!

JeffBezanson commented:
Looks good to me. I don't know if it matters, but `@horner` used to always use Horner's rule, while now it will sometimes use the more efficient algorithm for complex. Should it do what it says on the tin?

MasonProtter commented:
@JeffBezanson Good catch. One way I could fix that is by making it macroexpand to something like `invoke(evalpoly, Tuple{Any, Tuple}, x, p)`. Is that considered to be too ugly a thing to do here? I could also just write an inner generated function that always does Horner's rule.

JeffBezanson commented:
`invoke` is fine with me. We went to all that trouble to optimize it after all :)

MasonProtter committed:
Make @horner always use Horner's rule

MasonProtter commented:
Tests ran and passed on everything but `win32`, which timed out. Should I make a trivial commit to retrigger the tests, or is it fine?

MasonProtter committed:
fix docstring

musm commented:
wing 64 test failure unrelated, otherwise LGTM

musm commented:
@MasonProtter could you confirm that some of the elementary math function, e.g. `exp` and `exp10`  don't see regressions with the updated `@horner` macro?

MasonProtter commented:
I see only only performance gains, at least relative to v.13.0-rc3.
```julia
julia> using Printf

julia> begin
           @show VERSION
           for x in 10 .* randn(10)
               @show x
               print("  exp("*@sprintf("%.3f",x)*"):   "); @btime exp(  $(Ref(x))[])
               print("  exp10("*@sprintf("%.3f",x)*"): "); @btime exp10($(Ref(x))[])
           end
       end
VERSION = v"1.3.0-rc3.0"
x = -7.15243458415763
  exp(-7.152):     7.744 ns (0 allocations: 0 bytes)
  exp10(-7.152):   9.172 ns (0 allocations: 0 bytes)
x = -8.719494400795519
  exp(-8.719):     7.741 ns (0 allocations: 0 bytes)
  exp10(-8.719):   9.201 ns (0 allocations: 0 bytes)
x = -29.801956722722643
  exp(-29.802):     7.741 ns (0 allocations: 0 bytes)
  exp10(-29.802):   9.201 ns (0 allocations: 0 bytes)
x = -8.877111035670996
  exp(-8.877):     7.741 ns (0 allocations: 0 bytes)
  exp10(-8.877):   8.392 ns (0 allocations: 0 bytes)
x = -0.6928593337570201
  exp(-0.693):     6.449 ns (0 allocations: 0 bytes)
  exp10(-0.693):   9.200 ns (0 allocations: 0 bytes)
x = 2.866124556333614
  exp(2.866):     7.737 ns (0 allocations: 0 bytes)
  exp10(2.866):   8.117 ns (0 allocations: 0 bytes)
x = 23.658277524062772
  exp(23.658):     7.743 ns (0 allocations: 0 bytes)
  exp10(23.658):   9.196 ns (0 allocations: 0 bytes)
x = -8.531107335629256
  exp(-8.531):     7.748 ns (0 allocations: 0 bytes)
  exp10(-8.531):   9.203 ns (0 allocations: 0 bytes)
x = -1.907378473171837
  exp(-1.907):     7.748 ns (0 allocations: 0 bytes)
  exp10(-1.907):   9.199 ns (0 allocations: 0 bytes)
x = 12.353234804733408
  exp(12.353):     7.740 ns (0 allocations: 0 bytes)
  exp10(12.353):   9.198 ns (0 allocations: 0 bytes)
```
```
julia> using Printf

julia> begin
           @show VERSION
           for x in 10 .* randn(10)
               @show x
               print("  exp("*@sprintf("%.3f",x)*"):   "); @btime exp(  $(Ref(x))[])
               print("  exp10("*@sprintf("%.3f",x)*"): "); @btime exp10($(Ref(x))[])
           end
       end
VERSION = v"1.4.0-DEV.539"
x = 3.0797159237964333
  exp(3.080):     7.720 ns (0 allocations: 0 bytes)
  exp10(3.080):   8.107 ns (0 allocations: 0 bytes)
x = 11.524550980616235
  exp(11.525):     7.727 ns (0 allocations: 0 bytes)
  exp10(11.525):   8.647 ns (0 allocations: 0 bytes)
x = -13.828224826793532
  exp(-13.828):     7.726 ns (0 allocations: 0 bytes)
  exp10(-13.828):   8.652 ns (0 allocations: 0 bytes)
x = 10.171399187874462
  exp(10.171):     7.727 ns (0 allocations: 0 bytes)
  exp10(10.171):   8.648 ns (0 allocations: 0 bytes)
x = -14.692467227301847
  exp(-14.692):     7.719 ns (0 allocations: 0 bytes)
  exp10(-14.692):   8.650 ns (0 allocations: 0 bytes)
x = 12.429415466997678
  exp(12.429):     7.725 ns (0 allocations: 0 bytes)
  exp10(12.429):   8.646 ns (0 allocations: 0 bytes)
x = 3.5732507205285162
  exp(3.573):     7.727 ns (0 allocations: 0 bytes)
  exp10(3.573):   8.649 ns (0 allocations: 0 bytes)
x = -5.010189256550127
  exp(-5.010):     7.727 ns (0 allocations: 0 bytes)
  exp10(-5.010):   8.648 ns (0 allocations: 0 bytes)
x = -4.809001460910974
  exp(-4.809):     7.724 ns (0 allocations: 0 bytes)
  exp10(-4.809):   8.653 ns (0 allocations: 0 bytes)
x = 19.73389331115626
  exp(19.734):     7.724 ns (0 allocations: 0 bytes)
  exp10(19.734):   8.651 ns (0 allocations: 0 bytes)
```

musm commented:
Great! For 2.0 we should deprecate at-evalpoly. 

In the meantime, we can replace all instances of  `@horner` with `evalpoly` (there's only a handful of calls), and deprecate this unexported macro, since all of its usages are subsumed by the new `evalpoly` function.

chriselrod commented:
@oscardssmith 

This caused a regression for some functions using SIMD intrinsics with llvmcall, because the `evalpoly` function doesn't inline due to overestimated costs.
```julia
julia> @inline function expb_kernel(::Val{ℯ}, x::Union{Float32, SVec{W, Float32} where W})
           return SLEEFPirates.@horner(x, 1.0f0, 1.0f0, 0.5f0,
                               0.16666415f0, 0.041666083f0,
                               0.008375129f0, 0.0013956056f0)
       end
 expb_kernel (generic function with 1 method)

julia> @btime myexp($(Ref(sx32))[]) # vector
  3.512 ns (0 allocations: 0 bytes)
 SVec{16,Float32}<16639.281f0, 1.8628123f0, 5.235295f0, 7.906962f-5, 1822.3082f0, 2.4061686f-9, 0.0030119985f0, 0.0002081529f0, 7.7609496f0, 23.372593f0, 8.717485f11, 3.6029082f6, 13742.452f0, 3.1532597f-9, 0.007988205f0, 0.00020794437f0>

julia> @inline function expb_kernel(::Val{ℯ}, x::Union{Float32, SVec{W, Float32} where W})
           return evalpoly(x, (1.0f0, 1.0f0, 0.5f0,
                               0.16666415f0, 0.041666083f0,
                               0.008375129f0, 0.0013956056f0))
       end
 expb_kernel (generic function with 1 method)

julia> @btime myexp($(Ref(sx32))[]) # vector
  4.975 ns (0 allocations: 0 bytes)
 SVec{16,Float32}<16639.281f0, 1.8628123f0, 5.235295f0, 7.906962f-5, 1822.3082f0, 2.4061686f-9, 0.0030119985f0, 0.0002081529f0, 7.7609496f0, 23.372593f0, 8.717485f11, 3.6029082f6, 13742.452f0, 3.1532597f-9, 0.007988205f0, 0.00020794437f0>
```

KristofferC commented:
Ref https://github.com/JuliaLang/julia/pull/32753#issuecomment-517054433

MasonProtter commented:
Note that comment was in regards to the original implementation of this PR that relied on recursion, rather than a generated function.

stevengj commented:
Do we still need a generated function for this?

I noticed recently that the compiler seems able to unroll/inline `Base.Math._evalpoly(x, tuple)` (the loopy version).  I haven't found any case yet where the `@generated` version produces better code.

LilithHafner commented:
For code generation, no, but for effect inference, unfortunately, yes
```julia-repl
julia> Base.infer_effects(evalpoly, (Float64, Tuple{Float64, Float64, Float64}))
(+c,+e,+n,+t,+s)

julia> Base.infer_effects(Base.Math._evalpoly, (Float64, Tuple{Float64, Float64, Float64}))
(!c,+e,!n,!t,+s)
```
